### DISCUSSION TITLE
Pocketbase in custom build is not able to serve static files from `pb_public` dir

### DISCUSSION THREAD

**Discussion by proxima-b-alpha** [pocketbase/pocketbase#2102] [Q&A] [answered]
When I access `http://0.0.0.0:8080/favicon.png` or `http://0.0.0.0:8080/index.html` it throws `404`

Folder structure:

- test-pb
  - main.go
  - pb_data
  - pb_public
    - favicon.png
    - index.html

**main.go**

```
package main

import (
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/cmd"
)

func main() {
	app := pocketbase.New()
	app.Bootstrap()
	serveCmd := cmd.NewServeCommand(app, true)
	serveCmd.SetArgs([]string{"--http=0.0.0.0:8080"})
	serveCmd.Execute()
}
```

**Run application**

`CGO_ENABLED=0 go run main.go`

**Console output**

```
[0.00ms] SELECT {{_params}}.* FROM `_params` WHERE `key`='settings' LIMIT 1
[0.00ms] CREATE TABLE IF NOT EXISTS `_migrations` (file VARCHAR(255) PRIMARY KEY NOT NULL, applied INTEGER NOT NULL)
[0.00ms] SELECT count(*) FROM `_migrations` WHERE `file`='1640988000_init.go' LIMIT 1
[0.00ms] SELECT count(*) FROM `_migrations` WHERE `file`='1673167670_multi_match_migrate.go' LIMIT 1
[0.00ms] SELECT count(*) FROM `_migrations` WHERE `file`='1677152688_rename_authentik_to_oidc.go' LIMIT 1
[0.00ms] SELECT {{_params}}.* FROM `_params` WHERE `key`='settings' LIMIT 1
2023/03/20 21:52:31 Server started at http://0.0.0.0:8080
 ➜ REST API: http://0.0.0.0:8080/api/
 ➜ Admin UI: http://0.0.0.0:8080/_/
```

**Works with NOT custom build**

Same `pb_public` dir works totally fine when used with standalone pocketbase `./pocketbase.exe serve `

**Comment by ganigeorgiev** [comment, marked as answer]
You have to register the static file serve route manually. For an example, you can check the main file of the prebuilt executable - https://github.com/pocketbase/pocketbase/blob/master/examples/base/main.go:

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    // serves static files from the provided public dir (if exists)
    e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./public_dir"), false))
    return nil
})
```

**Reply by proxima-b-alpha** [reply]
Got it, Thank you 👍

**Comment by shubhendumadhukar** [comment]
This might not be a pocketbase specific question, but I am not able to serve some static pages generated by sveltekit. I usually serve my sveltekit site using Node.js/express like so:
```javascript
app.use(express.static("dist"))

app.all("*", (req, res) => {
    res.sendFile(path.resolve(path.join("dist", "index.html")))
})
```
I tried to replicate in pocketbase/echo, but following does not seem to work. I can access the root path, but if I refresh the page at a non-root path, I get a 404. Setting `indexFallback = true` helps me get rid of 404, but I get index.html for all routes (if I refresh the page)
```golang
//go:embed all:dist
var distDir embed.FS

func main() {
    app := pocketbase.New()
    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        var DistDirFS = echo.MustSubFS(distDir, "dist")
        e.Router.Any("/*", apis.StaticDirectoryHandler(DistDirFS, false))
        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Comment by loganmac** [comment]
@shubhendumadhukar I would guess it’s probably because you’re not rendering every file that the Go server would need to serve without node. Check out the guide for `adapter-static`: https://kit.svelte.dev/docs/adapter-static

**Reply by shubhendumadhukar** [reply]
My bad, I should have clarified, that I am not using the adapter-node. I was using express in a standalone project outside sveltekit. And yes, I did use static adapter to build out a dist folder with all the files images, html files..and it works fine if I use express to serve them or even the cli [serve](https://www.npmjs.com/package/serve).

EDIT: Problem was that my app isn't an SPA, so I needed to serve them differently. I noticed if I access the URLs with `.html`, they work just fine. I am sure there are better ways to do this, but I wrote a middleware to handle it, and it seems to be working now.
```golang
e.Router.Use(func(next echo.HandlerFunc) echo.HandlerFunc {
  return func(c echo.Context) error {
    if !strings.Contains(c.Request().URL.Path, ".") && !strings.HasSuffix(c.Request().URL.Path, "/") {
      c.Request().URL.Path += ".html"
    }
    return next(c)
  }
}, middleware.StaticWithConfig(middleware.StaticConfig{
    Filesystem: DistDirFS,
}))
```

**Reply by xaverine** [reply]
Svelte-Kit's server-side are nearly all node based, except for SSG. ([see adapters](https://kit.svelte.dev/docs/adapters))
Since  PB uses go as backend, you can only use Svelte (SPA), not Svelte-Kit (Fullstack).

**Reply by shubhendumadhukar** [reply]
Not necessarily. You can use sveltekit via static adapter as well. Can't use the sveltekit feature like ssr, prerender etc but if you don't need those, you can definitely use sveltekit generated build in pocketbase. Wails works on the same principles, tauri as well.

**Reply by xaverine** [reply]
correct me if I'm wrong!
svelte-kit via a static adapter, basically strip it down to..  well, a static site.
taking away most of its capabilities. 
what's the point in that?

**Reply by xaverine** [reply]
with pb_public:
I was originally planning to use it with Astro.
After some research, I've decided to go with SPA.
ps: seperate/standalone PB backends are not in this discussion.

**Reply by shubhendumadhukar** [reply]
For me it's primarily the routing and the sort of opinionated folder structure. But yes, don't really need to use sveltekit, svelte is fine instead.
